server/src/main/java/com/broadcom/lsp/cobol/service/CobolTextDocumentService.java

```diff
@@ -248,8 +265,13 @@ private void analyzeDocumentFirstTime(String uri, String text, boolean userReque
               ofNullable(docs.get(uri)).ifPresent(doc -> doc.setAnalysisResult(result));
               publishResult(uri, result, copybookProcessingMode);
               outlineMap.get(uri).complete(result.getOutlineTree());
-            })
-        .whenComplete(reportExceptionIfThrown(createDescriptiveErrorMessage("analysis", uri)));
+              completableFuture.complete(null);
+              return null;
+            });
+    futureMap.put(uri, docAnalysisFuture);
+    completableFuture
+        .whenComplete(reportExceptionIfThrown(createDescriptiveErrorMessage("analysis", uri)))
+        .whenComplete((val1, val2) -> clearAnalysedFutureObject(uri));
```

Comment:
There is no need for CompletableFuture at all. You can add these two calls at the end of the submitted function, instead of `completableFuture.complete(null);`.

Reply:
why didn't I think of this? 🤦